src/Zalo/Provider.php

```diff
@@ -20,52 +23,64 @@ class Provider extends AbstractProvider
      */
     protected function getAuthUrl($state)
     {
-        return $this->buildAuthUrlFromBase('https://oauth.zaloapp.com/v3/auth', $state);
+        return $this->buildAuthUrlFromBase('https://oauth.zaloapp.com/v4/permission', $state);
     }
 
     /**
      * {@inheritdoc}
      */
     protected function getCodeFields($state = null)
     {
-        $fields = [
-            'app_id'       => $this->clientId,
-            'redirect_uri' => $this->redirectUrl,
-            'state'        => $state,
-        ];
+        $fields = parent::getCodeFields($state);
 
-        return array_merge($fields, $this->parameters);
+        $fields['app_id'] = $this->clientId;
+
+        return $fields;
     }
 
     /**
      * {@inheritdoc}
      */
     public function getAccessTokenResponse($code)
     {
-        $response = $this->getHttpClient()->get($this->getTokenUrl(), [
-            RequestOptions::HEADERS => ['Accept' => 'application/json'],
-            RequestOptions::QUERY   => $this->getTokenFields($code),
+        $response = $this->getHttpClient()->post($this->getTokenUrl(), [
+            RequestOptions::HEADERS     => $this->getTokenHeaders($code),
+            RequestOptions::FORM_PARAMS => $this->getTokenFields($code),
         ]);
 
-        return json_decode((string) $response->getBody(), true);
+        return json_decode($response->getBody(), true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenHeaders($code)
+    {
+        return [
+            'Content-Type' => 'application/x-www-form-urlencoded',
+            'secret_key'   => $this->clientSecret,
```

Comment:
As this is the only header can you just set it in the get access token response method?